Source/EasyNetQ/RabbitAdvancedBus.cs

```diff
@@ -608,49 +597,52 @@ public async Task ExchangeDeclarePassiveAsync(string name, CancellationToken can
-        public virtual async Task UnbindAsync(IBinding binding, CancellationToken cancellationToken)
+        public virtual async Task UnbindAsync(Binding<Queue> binding, CancellationToken cancellationToken)
         {
-            Preconditions.CheckNotNull(binding, "binding");
-
             using var cts = cancellationToken.WithTimeout(configuration.Timeout);
 
-            if (binding.Bindable is IQueue queue)
-            {
-                await clientCommandDispatcher.InvokeAsync(
-                    x => x.QueueUnbind(queue.Name, binding.Exchange.Name, binding.RoutingKey, null),
-                    cts.Token
-                ).ConfigureAwait(false);
+            await clientCommandDispatcher.InvokeAsync(
+                x => x.QueueUnbind(
+                    binding.Destination.Name, binding.Source.Name, binding.RoutingKey, binding.Arguments?.ToDictionary(e => e.Key, e => e.Value)
+                ),
+                cts.Token
+            ).ConfigureAwait(false);
 
-                if (logger.IsDebugEnabled())
-                {
-                    logger.DebugFormat(
-                        "Unbound queue {queue} from exchange {exchange} with routing key {routingKey}",
-                        queue.Name,
-                        binding.Exchange.Name,
-                        binding.RoutingKey
-                    );
-                }
-            }
-            else if (binding.Bindable is IExchange destination)
+            if (logger.IsDebugEnabled())
             {
-                await clientCommandDispatcher.InvokeAsync(
-                    x => x.ExchangeUnbind(destination.Name, binding.Exchange.Name, binding.RoutingKey, null),
-                    cts.Token
-                ).ConfigureAwait(false);
+                logger.DebugFormat(
+                    "Unbound queue {queue} from exchange {exchange} with routing key {routingKey}",
+                    binding.Destination.Name,
+                    binding.Source.Name,
+                    binding.RoutingKey
+                );
+            }
+        }
 
-                if (logger.IsDebugEnabled())
-                {
-                    logger.DebugFormat(
-                        "Unbound destination exchange {destinationExchange} from source exchange {sourceExchange} with routing key {routingKey}",
-                        destination.Name,
-                        binding.Exchange.Name,
-                        binding.RoutingKey
-                    );
-                }
+        /// <inheritdoc />
+        public virtual async Task UnbindAsync(Binding<Exchange> binding, CancellationToken cancellationToken)
+        {
+            using var cts = cancellationToken.WithTimeout(configuration.Timeout);
+
+            await clientCommandDispatcher.InvokeAsync(
+                x => x.ExchangeUnbind(
+                    binding.Destination.Name, binding.Source.Name, binding.RoutingKey, binding.Arguments?.ToDictionary(e => e.Key, e => e.Value)
```

Comment:
`binding.Arguments?.ToDictionary(e => e.Key, e => e.Value)` 

We have to convert IReadOnlyDictionary to IDictionary

Reply:
otherwise we need also to change `ExchangeUnbind` isn't it?

Reply:
doing also the change on `ExchangeUnbind` will it be a PITA?

Reply:
Can you elaborate? ExchangeUnbind is `RabbitMQ.Client` method and it only accepts IDictionary 😞

Reply:
> Can you elaborate? ExchangeUnbind is `RabbitMQ.Client` method and it only accepts IDictionary 😞

ops 😄 so that's ok what we have done or leave IDictionary also our side

Reply:
>leave IDictionary also our side

I have the same thoughts... it looks like we have two consistent options: 1) leave it as is 2) change `Arguments` from `IReadOnlyDictionary` to `IDictionary` for `Queue`, `Exchange` and `Binding`.

I don't know what to pick, so feel free to help 😉

Reply:
I'm for IDictionary.

Reply:
me too, you would need more code to manage it, and more instantiations probably for nothing back

Reply:
Done